core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -639,6 +637,11 @@ class SparkContext(config: SparkConf) extends Logging {
     postEnvironmentUpdate()
     postApplicationStart()
 
+    // After application started, attach handlers to started server and start handler.
+    _ui.foreach(_.attachAllHandler())
```

Comment:
btw, why do we need to attach all the handlers again?

Reply:
> btw, why do we need to attach all the handlers again? And there are so many "attachAllHandler" in the code changes.

That me clarify the process.
Before this change. We add handlers to SparkUI but not start it when init SparkUI since we not start jetty server.
Then it call `bind()`, in this method we start jetty server and attach all handler to server.
this cause a problem that when we call `bind()`, we expose all servlet API to user. But application not fully started yet.
Now in this pr, I split the behavior of start jetty server and attach handlers to server.
We need to bind address(start jetty server) first before start AM since we need driver url address to bind with spark proxy server.

Then after application fully started, we attach all handlers to server(means expose UI url to user).


For this comment https://github.com/apache/spark/pull/33457#issuecomment-886624148
I add a initHandler to handle all request between  starting jetty server and  fully start application to show the hint message.

Reply:
> I split the behavior of start jetty server and attach handlers to server

I add some logging and find that `attachHandler` is still called multiple times before `attachAllHandler`

Reply:
> > I split the behavior of start jetty server and attach handlers to server
> 
> I add some logging and find that `attachHandler` is still called multiple times before `attachAllHandler`

Yes, but when we call `attachHandler()` the `serverInfo` is none, so the handler is not attached to the server.

Reply:
Is there a way to avoid calling `attachAllHandler` in so many places?

Reply:
> Is there a way to avoid calling `attachAllHandler` in so many places?

Add a new method call it `bindAndAttachAllHandler()` then replace other place to call this one? only call `bind()` and `attachAllHandler` in  SparkContext?

Reply:
Will that work for history server and master/worker UI? If yes let's try to avoid changing them.

Reply:
> Will that work for history server and master/worker UI? If yes let's try to avoid changing them.

How about current?

Reply:
I mean, will history server and master/worker UI show the same page on starting up?

Reply:
> I mean, will history server and master/worker UI show the same page(spark is starting up...) on starting up?

With current code, won't impact history server/master/worker web UI. and   history server/master/worker web UI doesn't have such problem.